src/CustomElementRegistry.js

```diff
@@ -22,55 +21,23 @@ export default class CustomElementRegistry {
    * @param {!CustomElementInternals} internals
    */
   constructor(internals) {
-    /**
-     * @private
-     * @type {boolean}
-     */
-    this._elementDefinitionIsRunning = false;
 
     /**
      * @private
      * @type {!CustomElementInternals}
      */
     this._internals = internals;
 
-    /**
-     * @private
-     * @type {!Map<string, !Deferred<undefined>>}
-     */
-    this._whenDefinedDeferred = new Map();
-
-    /**
-     * The default flush callback triggers the document walk synchronously.
-     * @private
-     * @type {!Function}
-     */
-    this._flushCallback = fn => fn();
-
-    /**
-     * @private
-     * @type {boolean}
-     */
-    this._flushPending = false;
-
-    /**
-     * @private
-     * @type {!Array<!CustomElementDefinition>}
-     */
-    this._pendingDefinitions = [];
-
-    /**
-     * @private
-     * @type {!DocumentConstructionObserver}
-     */
-    this._documentConstructionObserver = new DocumentConstructionObserver(internals, document);
+    if (!internals.preferPerformance) {
+      /**
+      * @private
+      * @type {!DocumentConstructionObserver}
+      */
+      this._documentConstructionObserver = new DocumentConstructionObserver(internals, document);
+    }
   }
 
-  /**
-   * @param {string} localName
-   * @param {!Function} constructor
-   */
-  define(localName, constructor) {
+  _validateDefinition(localName, constructor) {
```

Comment:
I think we should consider renaming this because it's not really validating an existing definition. It's doing something more like checking if you're allowed to create a definition using the given arguments at the current time - `_assertCanDefine`?

Reply:
Done.